src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1284 @@
+            {
+                TryPushOperand(NegateIfNeeded(node));
+                return null;
+            }
+
+            public override BoundNode? VisitConstantPattern(BoundConstantPattern node)
+            {
+                TryPushOperand(NegateIfNeeded(node));
+                return null;
+            }
+
+            public override BoundNode? VisitDiscardPattern(BoundDiscardPattern node)
+            {
+                TryPushOperand(NegateIfNeeded(node));
+                return null;
+            }
+
+            public override BoundNode? VisitRelationalPattern(BoundRelationalPattern node)
+            {
+                TryPushOperand(NegateIfNeeded(node));
+                return null;
+            }
+
+            // Given a pattern that expects `pattern.InputType` as input type, produce
+            // one that expects `inputType.
+            //
+            // For example, when splitting a pattern `{ Prop: A and B }` into `{ Prop: A } and/or { Prop: B }`
+            // the original `B` pattern expects an input type that is the narrowed type from `A`
+            // but the rewritten `B` pattern expects an input type that is the type of `Prop`.
+            private static BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern, TypeSymbol inputType)
+            {
+                if (pattern.InputType.Equals(inputType, TypeCompareKind.AllIgnoreOptions))
+                {
+                    return pattern;
+                }
+
+                if (pattern is BoundTypePattern typePattern1)
+                {
+                    return typePattern1.Update(typePattern1.DeclaredType, typePattern1.IsExplicitNotNullTest, inputType, typePattern1.NarrowedType);
+                }
+
+                if (pattern is BoundRecursivePattern recursivePattern)
+                {
+                    return recursivePattern.Update(
+                        recursivePattern.DeclaredType ??
+                            new BoundTypeExpression(recursivePattern.Syntax, aliasOpt: null, recursivePattern.InputType.StrippedType()),
+                        recursivePattern.DeconstructMethod, recursivePattern.Deconstruction,
+                        recursivePattern.Properties, recursivePattern.IsExplicitNotNullTest,
+                        recursivePattern.Variable, recursivePattern.VariableAccess,
+                        inputType, recursivePattern.NarrowedType);
+                }
+
+                if (pattern is BoundDiscardPattern discardPattern)
+                {
+                    return discardPattern.Update(inputType, inputType);
+                }
+
+                if (pattern is BoundNegatedPattern negatedPattern)
+                {
+                    return negatedPattern.Update(
+                        WithInputTypeCheckIfNeeded(negatedPattern.Negated, inputType),
+                        inputType, inputType);
+                }
+
+                Debug.Assert(pattern is not BoundBinaryPattern);
```

Comment:
>Debug.Assert(pattern is not BoundBinaryPattern); [](http://example.com/codeflow?start=16&length=48)

It would be good to adjust this assert to enumerate specific kinds of patterns that we expect to see here. Are constant, relational, ITuple and list patterns the only left?